### PR TITLE
3554 Collection resource performance improvements

### PR DESCRIPTION
**Pull request by pkdash**
### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- [x] Automated Testing

- [x] Sufficient User and Developer Documentation

- [x] Passing Jenkins Build

- [x] Peer Code review and approval

### Positive Test Case
1. In order to see the performance improvements in terms of page load time for collection resource, you need to have at least 100 resources that can be added to a collection resource. Locally may be you can write a script to create that many composite resources. 
2. Checkout the develop branch.  Create at least 100 composite resources. Create a collection resource. Note down the load time  for the collection resource in edit mode. Try to add credit metadata for the collection resource and note down the page reload time.
3. Checkout the branch '3554-collection-resource-speedup' and compare the page load time for the collection resource with respective load times that you observed in step 2 above. You should see much better page load time compared to the 'develop' branch. 

**NOTE**: This work has been deployed on cuahsi-dev-2.hydroshare.org. You can use this site to verify collection resource page load time improvement compared to production or beta.

### CODE REVIEW

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4616](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4616)

File: hs_collection_resource/receivers.py [hunk 1]
```diff
@@ -21,13 +21,11 @@ def pre_check_bag_flag_handler(sender, **kwargs):
     if collection_res_obj.update_text_file.lower() == 'true':
         update_collection_list_csv(collection_res_obj)
         set_dirty_bag_flag(collection_res_obj)
-        collection_res_obj.extra_data = {'update_text_file': 'False'}
-        collection_res_obj.save()
+        collection_res_obj.set_update_text_file(flag='True')
```
Comment (devincowan) [inline hs_collection_resource/receivers.py:24]: Shouldn't we update the flag to 'False' in this case?
```suggestion
        collection_res_obj.set_update_text_file(flag='False')
```

**Review by devincowan — COMMENTED** [review]
Seems like a great improvement on page load time--nice!
I left a few comments that might need explaining

File: theme/templates/resource-landing-page/coverage.html [hunk 2]
```diff
@@ -251,13 +251,15 @@ <h4>Spatial:</h4>
...
                                 {# we will show/hide this button from JS code when ever coverage element gets updated for aggregation #}
-                                <button id="btn-update-resource-spatial-coverage"
-                                        type="button" class="btn btn-info space-top"
-                                        {% if not cm.resource_type == "CompositeResource" or not cm.has_logical_spatial_coverage %}
-                                        style="display: none"
-                                        {% endif %}>
-                                    <i class="fa fa-calculator" aria-hidden="true"></i> Set spatial coverage from content files
-                                </button>
+                                {% if cm.resource_type == "CompositeResource" %}
+                                    <button id="btn-update-resource-spatial-coverage"
```
Comment (devincowan) [inline theme/templates/resource-landing-page/coverage.html:255]: I did notice that the buttons to "Set [spatial/temporal] coverage from content files APPEAR when one starts typing in the related input fields--_regardless of whether the input is "logical" or not._ It seems odd to offer to set coverage while the user it actively typing in a coverage manually. But this is not an issue particular to this PR

Comment (pkdash) [reply theme/templates/resource-landing-page/coverage.html:255]: There are other issues with that button - it should not be displayed when there is no aggregation with spatial coverage. I will create a sperate issue related to this button.

File: hydroshare/settings.py [hunk 3]
```diff
@@ -774,3 +774,5 @@
 #sys.stderr = codecs.getwriter('utf8')(sys.stderr)
 
 MODEL_PROGRAM_META_SCHEMA_TEMPLATE_PATH = "/hydroshare/hs_file_types/model_meta_schema_templates"
+# Number of resources to list when adding resources to a collection resource
+COLLECTABLE_RESOURCES_PAGE_SIZE = 50
```
[the remark this answers is not in the document]
Comment (pkdash) [reply hydroshare/settings.py]: Fixed

File: hs_collection_resource/views.py [hunk 4]
```diff
@@ -3,16 +3,52 @@
...
+    status_code = http_status.HTTP_200_OK
+    try:
+        collection_res, is_authorized, user = authorize(request, shortkey,
+                                                        needed_permission=ACTION_TO_AUTHORIZE.EDIT_RESOURCE)
+
+        if collection_res.resource_type.lower() != "collectionresource":
+            raise Exception(f"Resource {shortkey} is not a collection resource.")
+
+        if collection_res.raccess.published:
+            raise Exception(f"Resource {shortkey} of a published collection resource and can't be changed")
```
[the remark this answers is not in the document]
Comment (pkdash) [reply hs_collection_resource/views.py]: Fixed

File: hs_collection_resource/receivers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pkdash) [reply hs_collection_resource/receivers.py]: Thanks for catching this. Fixed.

File: hs_collection_resource/templates/pages/collectable_resources_modal.html [hunk 5]
```diff
@@ -0,0 +1,94 @@
+{% load hydroshare_tags %}
+<div class="modal fade" id="collection-candidate" role="dialog" aria-labelledby="myModalLabel" aria-hidden="true">
+    <div class="modal-dialog" style="width:65%;">
+        <div class="modal-content">
+            <div class="modal-header">
+                <button type="button" class="close" data-dismiss="modal" aria-hidden="true">&times;</button>
+                <h4 class="modal-title">Collectable Resources</h4>
+            </div>
+        <div class="modal-body">
```
[the remark this answers is not in the document]
Comment (pkdash) [reply hs_collection_resource/templates/pages/collectable_resources_modal.html]: Fixed.

File: theme/templates/resource-landing-page/coverage.html (hunk 2 above)
Comment (devincowan) [reply theme/templates/resource-landing-page/coverage.html:255]: sounds great, thanks!

**Review by devincowan — APPROVED** [review]
Looks great, thanks Pabitra. I think this is going to make Collections vastly more usable!

**Pull request merged**